File: python/runpack/runner.py
"""Main runner loop for executing Runpack jobs."""

import logging
import signal
import time
from typing import Optional

from .client import RunpackClient
from .config import (
    MIN_POLL_INTERVAL,
    MAX_POLL_INTERVAL,
    POLL_INTERVAL_INCREMENT,
    WORKER_URL,
    RunnerConfig,
)
from .jobs import get_job_handler, get_supported_job_types


logger = logging.getLogger(__name__)


class JobRunner:
    """Main job runner that polls for and executes jobs."""
    
    def __init__(self, api_key: str):
        """Initialize the job runner.
        
        Args:
            api_key: Runner API key for authentication
        """
        self.api_key = api_key
        self.client = RunpackClient(WORKER_URL, api_key)
        self.config = RunnerConfig()
        self.running = True
        self.current_job_id: Optional[str] = None
        self.current_poll_interval = MIN_POLL_INTERVAL
        
        # Setup signal handlers for graceful shutdown
        signal.signal(signal.SIGINT, self._signal_handler)
        signal.signal(signal.SIGTERM, self._signal_handler)
    
    def _signal_handler(self, signum, frame):
        """Handle shutdown signals gracefully."""
        logger.info(f"Received signal {signum}, shutting down gracefully...")
        self.running = False
    
    def register(self) -> None:
        """Register the runner with the worker or load existing registration."""
        # Try to load existing configuration
        if self.config.load():
            logger.info(f"Loaded existing runner configuration: {self.config.runner_name}")
            logger.info(f"Runner ID: {self.config.runner_id}")
            self.client.runner_id = self.config.runner_id
            
            # Verify that the runner ID is still valid in the system
            logger.info("Verifying runner registration...")
            try:
                is_valid = self.client.verify_runner()
                if not is_valid:
                    raise RuntimeError(
                        f"Runner ID '{self.config.runner_id}' is not registered in the system.\n"
                        f"This can happen if the runner was deleted from the database.\n"
                        f"To fix this, delete the config file: {self.config.config_file}\n"
                        f"Then restart the runner to register with a new ID."
                    )
                logger.info("Runner registration verified successfully")
            except Exception as e:
                if isinstance(e, RuntimeError):
                    # Re-raise our custom error
                    raise
                else:
                    # Network or other error during verification
                    logger.error(f"Failed to verify runner registration: {e}")
                    raise RuntimeError(
                        f"Failed to verify runner registration: {e}\n"
                        f"Check your network connection and worker URL."
                    )
        else:
            # Generate new runner name and register
            runner_name = self.config.generate_name()
            capabilities = get_supported_job_types()
            
            runner_id = self.client.register_runner(runner_name, capabilities)
            self.config.set_runner_info(runner_id, runner_name)
            
            logger.info(f"Registered new runner: {runner_name}")
            logger.info(f"Runner ID: {runner_id}")
            logger.info(f"Capabilities: {capabilities}")
    
    def execute_job(self, job_id: str, job_type: str, input_params: dict) -> None:
        """Execute a job.
        
        Args:
            job_id: Job ID
            job_type: Type of job to execute
            input_params: Job input parameters
        """
        self.current_job_id = job_id
        logger.info(f"Executing job {job_id} of type '{job_type}'")
        
        try:
            # Get the job handler
            handler_class = get_job_handler(job_type)
            handler = handler_class()
            
            # Create heartbeat callback
            def send_heartbeat(progress_current: int, progress_total: int, console_output: str):
                """Send heartbeat to worker."""
                try:
                    self.client.send_heartbeat(
                        job_id=job_id,
                        progress_current=progress_current,
                        progress_total=progress_total,
                        console_output=console_output
                    )
                    logger.debug(f"Sent heartbeat for job {job_id}: {progress_current}/{progress_total}")
                except Exception as e:
                    logger.error(f"Failed to send heartbeat: {e}")
            
            # Execute the job
            result = handler.execute(input_params, send_heartbeat)
            
            # Extract console output from result if present
            console_output = result.pop('console_output', '')
            
            # Report successful completion
            self.client.complete_job(
                job_id=job_id,
                output_data=result,
                console_output=console_output
            )
            
            logger.info(f"Successfully completed job {job_id}")
            
        except ValueError as e:
            # Parameter validation error
            error_msg = f"Invalid parameters: {str(e)}"
            logger.error(f"Job {job_id} failed: {error_msg}")
            
            self.client.error_job(
                job_id=job_id,
                error_message=error_msg,
                console_output=error_msg
            )
            
        except Exception as e:
            # Unexpected error during execution
            error_msg = f"Job execution failed: {str(e)}"
            logger.error(f"Job {job_id} failed with exception: {e}", exc_info=True)
            
            self.client.error_job(
                job_id=job_id,
                error_message=error_msg,
                console_output=error_msg
            )
        
        finally:
            self.current_job_id = None
    
    def poll_and_execute(self) -> bool:
        """Poll for available jobs and execute them.
        
        Returns:
            True if a job was executed, False otherwise
        """
        try:
            # Get supported job types
            job_types = get_supported_job_types()
            
            # Poll for available jobs
            available_jobs = self.client.get_available_jobs(job_types)

            if not available_jobs:
                logger.debug("No jobs available")
                return False
            
            # Try to claim the first available job
            for job in available_jobs:
                job_id = job['job_id']
                
                claimed_job = self.client.claim_job(job_id)
                
                if claimed_job:
                    # Successfully claimed, execute it
                    self.execute_job(
                        job_id=claimed_job['job_id'],
                        job_type=claimed_job['job_type'],
                        input_params=claimed_job['input_params']
                    )
                    return True  # Job was executed
                else:
                    logger.debug(f"Failed to claim job {job_id}, trying next one")
            
            # No jobs were successfully claimed
            return False
        
        except Exception as e:
            logger.error(f"Error during polling: {e}", exc_info=True)
            return False
    
    def run(self) -> None:
        """Main runner loop."""
        logger.info("Starting Runpack job runner")
        logger.info(f"Worker URL: {WORKER_URL}")
        logger.info(f"Poll interval: {MIN_POLL_INTERVAL}-{MAX_POLL_INTERVAL} seconds (progressive backoff)")
        
        # Register the runner
        self.register()
        
        # Main loop
        logger.info("Entering main polling loop...")
        logger.info(f"Initial poll interval: {self.current_poll_interval} seconds")
        
        while self.running:
            try:
                job_executed = self.poll_and_execute()
                
                # Adjust polling interval based on whether a job was executed
                if job_executed:
                    # Job was executed - reset to minimum interval
                    if self.current_poll_interval != MIN_POLL_INTERVAL:
                        self.current_poll_interval = MIN_POLL_INTERVAL
                        logger.info(f"Job executed - resetting poll interval to {MIN_POLL_INTERVAL} seconds")
                else:
                    # No job executed - increase interval (with max cap)
                    old_interval = self.current_poll_interval
                    self.current_poll_interval = min(
                        self.current_poll_interval + POLL_INTERVAL_INCREMENT,
                        MAX_POLL_INTERVAL
                    )
                    if self.current_poll_interval != old_interval:
                        logger.info(f"No jobs available - increasing poll interval to {self.current_poll_interval} seconds")
                
            except KeyboardInterrupt:
                # Handled by signal handler
                break
            except Exception as e:
                logger.error(f"Unexpected error in main loop: {e}", exc_info=True)
            
            # Sleep before next poll, but check for shutdown frequently
            if self.running:
                # Sleep in 1-second intervals to allow for responsive shutdown
                for _ in range(self.current_poll_interval):
                    if not self.running:
                        break
                    time.sleep(1)
        
        # Shutdown
        if self.current_job_id:
            logger.warning(f"Shutting down with job {self.current_job_id} still in progress")
        
        logger.info("Runner stopped")


def run_runner(api_key: str) -> None:
    """Run the job runner.
    
    Args:
        api_key: Runner API key
    """
    runner = JobRunner(api_key)
    runner.run()

File: worker/src/handlers/runner.ts
import {
  claimJob,
  completeJob,
  failJob,
  getAvailableJobs,
  getJobById,
  getRunnerById,
  registerRunner,
  updateJobHeartbeat,
  updateRunnerLastSeen,
} from '../db/queries';
import {
  AvailableJob,
  ClaimJobResponse,
  CompleteJobRequest,
  CompleteJobResponse,
  Env,
  ErrorJobRequest,
  ErrorJobResponse,
  HeartbeatRequest,
  HeartbeatResponse,
  RegisterRunnerRequest,
  RegisterRunnerResponse
} from '../types';
import { generateId } from '../utils/hash';
import { validateConsoleOutput, validateErrorMessage, validateJobOutput } from '../utils/validation';

/**
 * Handle runner verification
 */
export async function handleVerifyRunner(request: Request, env: Env, runnerId: string): Promise<Response> {
  try {
    // Check if runner exists in database
    const runner = await getRunnerById(env, runnerId);

    if (!runner) {
      return new Response(JSON.stringify({
        exists: false,
        message: 'Runner not found',
      }), {
        status: 404,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    return new Response(JSON.stringify({
      exists: true,
      runner_name: runner.name,
      message: 'Runner verified successfully',
    }), {
      status: 200,
      headers: { 'Content-Type': 'application/json' },
    });
  } catch (error) {
    return new Response(JSON.stringify({
      error: 'Failed to verify runner',
      details: error instanceof Error ? error.message : 'Unknown error'
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });
  }
}

/**
 * Handle runner registration
 */
export async function handleRegisterRunner(request: Request, env: Env): Promise<Response> {
  try {
    const body = await request.json() as RegisterRunnerRequest;
    const { name, capabilities } = body;

    if (!name || !capabilities || !Array.isArray(capabilities)) {
      return new Response(JSON.stringify({ error: 'Invalid registration data' }), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    // Generate runner ID
    const runnerId = generateId();

    // Register runner in database
    await registerRunner(env, runnerId, name, JSON.stringify(capabilities));

    const response: RegisterRunnerResponse = {
      runner_id: runnerId,
      message: 'Runner registered successfully',
    };

    return new Response(JSON.stringify(response), {
      status: 201,
      headers: { 'Content-Type': 'application/json' },
    });
  } catch (error) {
    return new Response(JSON.stringify({
      error: 'Failed to register runner',
      details: error instanceof Error ? error.message : 'Unknown error'
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });
  }
}

/**
 * Handle get available jobs
 */
export async function handleGetAvailableJobs(request: Request, env: Env, runnerId: string): Promise<Response> {
  try {
    // Update runner last seen
    await updateRunnerLastSeen(env, runnerId);

    // Parse query parameters
    const url = new URL(request.url);
    const types = url.searchParams.getAll('types[]');

    if (types.length === 0) {
      return new Response(JSON.stringify({ jobs: [] }), {
        status: 200,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    // Get available jobs
    const jobs = await getAvailableJobs(env, types);

    const availableJobs: AvailableJob[] = jobs.map(job => ({
      job_id: job.id,
      job_type: job.job_type,
      input_params: JSON.parse(job.input_params),
      created_at: job.created_at,
    }));

    return new Response(JSON.stringify({ jobs: availableJobs }), {
      status: 200,
      headers: { 'Content-Type': 'application/json' },
    });
  } catch (error) {
    return new Response(JSON.stringify({
      error: 'Failed to get available jobs',
      details: error instanceof Error ? error.message : 'Unknown error'
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });
  }
}

/**
 * Handle claim job
 */
export async function handleClaimJob(request: Request, env: Env, jobId: string, runnerId: string): Promise<Response> {
  try {
    // Update runner last seen
    await updateRunnerLastSeen(env, runnerId);

    // Try to claim the job
    const result = await claimJob(env, jobId, runnerId);

    if (!result.success) {
      const response: ClaimJobResponse = {
        success: false,
        message: result.error || 'Failed to claim job',
      };
      return new Response(JSON.stringify(response), {
        status: 409,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    // Get job details
    const job = await getJobById(env, jobId);
    if (!job) {
      const response: ClaimJobResponse = {
        success: false,
        message: 'Job not found',
      };
      return new Response(JSON.stringify(response), {
        status: 404,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    const response: ClaimJobResponse = {
      success: true,
      message: 'Job claimed successfully',
      job: {
        job_id: job.id,
        job_type: job.job_type,
        input_params: JSON.parse(job.input_params),
      },
    };

    return new Response(JSON.stringify(response), {
      status: 200,
      headers: { 'Content-Type': 'application/json' },
    });
  } catch (error) {
    console.error('Error in handleClaimJob:', error);
    return new Response(JSON.stringify({
      error: 'Failed to claim job',
      details: error instanceof Error ? error.message : 'Unknown error'
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });
  }
}

/**
 * Handle heartbeat
 */
export async function handleHeartbeat(request: Request, env: Env, jobId: string, runnerId: string): Promise<Response> {
  try {
    const body = await request.json() as HeartbeatRequest;
    const { progress_current, progress_total, console_output } = body;

    // Validate console output size
    const validation = validateConsoleOutput(console_output);
    if (!validation.valid) {
      return new Response(JSON.stringify({ error: validation.error }), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    // Update runner last seen
    await updateRunnerLastSeen(env, runnerId);

    // Update job heartbeat
    const result = await updateJobHeartbeat(env, jobId, runnerId, progress_current, progress_total, console_output);

    if (!result.success) {
      const response: HeartbeatResponse = {
        success: false,
        message: result.error || 'Failed to update heartbeat',
      };
      return new Response(JSON.stringify(response), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    const response: HeartbeatResponse = {
      success: true,
      message: 'Heartbeat updated successfully',
    };

    return new Response(JSON.stringify(response), {
      status: 200,
      headers: { 'Content-Type': 'application/json' },
    });
  } catch (error) {
    return new Response(JSON.stringify({
      error: 'Failed to update heartbeat',
      details: error instanceof Error ? error.message : 'Unknown error'
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });
  }
}

/**
 * Handle complete job
 */
export async function handleCompleteJob(request: Request, env: Env, jobId: string, runnerId: string): Promise<Response> {
  try {
    const body = await request.json() as CompleteJobRequest;
    const { output_data, console_output } = body;

    // Validate output
    const outputValidation = validateJobOutput(output_data);
    if (!outputValidation.valid) {
      return new Response(JSON.stringify({ error: outputValidation.error }), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    const consoleValidation = validateConsoleOutput(console_output);
    if (!consoleValidation.valid) {
      return new Response(JSON.stringify({ error: consoleValidation.error }), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    // Update runner last seen
    await updateRunnerLastSeen(env, runnerId);

    // Complete the job
    const result = await completeJob(env, jobId, runnerId, JSON.stringify(output_data), console_output);

    if (!result.success) {
      const response: CompleteJobResponse = {
        success: false,
        message: result.error || 'Failed to complete job',
      };
      return new Response(JSON.stringify(response), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    const response: CompleteJobResponse = {
      success: true,
      message: 'Job completed successfully',
    };

    return new Response(JSON.stringify(response), {
      status: 200,
      headers: { 'Content-Type': 'application/json' },
    });
  } catch (error) {
    return new Response(JSON.stringify({
      error: 'Failed to complete job',
      details: error instanceof Error ? error.message : 'Unknown error'
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });
  }
}

/**
 * Handle error job
 */
export async function handleErrorJob(request: Request, env: Env, jobId: string, runnerId: string): Promise<Response> {
  try {
    const body = await request.json() as ErrorJobRequest;
    const { error_message, console_output } = body;

    // Validate error message and console output
    const errorValidation = validateErrorMessage(error_message);
    if (!errorValidation.valid) {
      return new Response(JSON.stringify({ error: errorValidation.error }), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    const consoleValidation = validateConsoleOutput(console_output);
    if (!consoleValidation.valid) {
      return new Response(JSON.stringify({ error: consoleValidation.error }), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    // Update runner last seen
    await updateRunnerLastSeen(env, runnerId);

    // Fail the job
    const result = await failJob(env, jobId, runnerId, error_message, console_output);

    if (!result.success) {
      const response: ErrorJobResponse = {
        success: false,
        message: result.error || 'Failed to mark job as failed',
      };
      return new Response(JSON.stringify(response), {
        status: 400,
        headers: { 'Content-Type': 'application/json' },
      });
    }

    const response: ErrorJobResponse = {
      success: true,
      message: 'Job marked as failed successfully',
    };

    return new Response(JSON.stringify(response), {
      status: 200,
      headers: { 'Content-Type': 'application/json' },
    });
  } catch (error) {
    return new Response(JSON.stringify({
      error: 'Failed to mark job as failed',
      details: error instanceof Error ? error.message : 'Unknown error'
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });
  }
}
